benchmark/benchmark.py

```diff
@@ -0,0 +1,117 @@
+"""Benchmarking utility for PennyLane.
+"""
+# pylint: disable=import-outside-toplevel,invalid-name
+import functools
+import importlib
+import subprocess
+import sys
+
+import numpy as np
+
+import pennylane as qml
+
+
+def timing(func, *, number=10, repeat=5):
+    """Time the given function.
+
+    Args:
+        func (callable[None, Any]): function to time
+        number (int): number of loops per test
+        repeat (int): the number the timing test is run
+    """
+    import timeit
+
+    print('{} loops, {} runs'.format(number, repeat))
+    res = timeit.repeat(func, number=number, repeat=repeat, globals=globals())
+    print("Timing per loop:", np.array(res) / number)
+
+
+def plot(title, kernels, n_range):
+    """Plot timings as a function of the parameter n.
+
+    Args:
+        title (str): plot title
+        kernels (Sequence[callable[Any, Any]]): parametrized benchmarks to time
+        n_range (Sequence[Any]): values the benchmark parameter n takes
+    """
+    import perfplot
+
+    labels = [k.__doc__ for k in kernels]
+    perfplot.show(
+        setup=lambda n: n,
+        kernels=kernels,
+        labels=labels,
+        n_range=n_range,
+        xlabel="n",
+        title=title,
+        # logx="auto",  # set to True or False to force scaling
```

Comment:
What's the plan for these commented-out bits?

Reply:
Perfplot is a third-party module from PyPI, https://pypi.org/project/perfplot/. It'd be nice if it were just an optional dependency, not a required one.

It plots the time the various kernels (here benchmarks) take to execute, as a function of a scalar parameter `n` (which usually describes the size of the problem). The commented-out parts adjust the properties of the plot. I'll need to play with them a bit to see which options produce the most useful-looking plot.